src/platforms/common/blackpill-f4/blackpill-f4.c

```diff
@@ -194,3 +195,8 @@ uint8_t platform_spi_xfer(const spi_bus_e bus, const uint8_t value)
 	(void)bus;
 	return value;
 }
+
+int platform_hwversion(void)
```

Comment:
Is there a particular reason for removing this from the platform header? This will impair the optimiser's ability to optimise this function out to just a value in all call sites

Reply:
For `blackpill-f4` "all call sites" boil down to `cmd_version()` in src/command.c and `jtagtap_reset()`. First one prints board revision to gdb, second manipulates jTRST pin if available (and incurs a 10000 loops delay anyway). Flash size is not affected by this diff hunk.

Reply:
We aren't asking for speed reasons, we're asking for reasons of code size and generated code complexity and the ability for the optimiser to potentially just entirely optimise things away as a result.

Reply:
Code size is not affected significantly here. Complexity is reduced in my opinion as now I can clearly see where the function is actually used for a given platform by checking the callers in Puncover. LTO is the ultimate optimizer, but we're not going there yet.

Reply:
You can find all call sites without doing this change by using an IDE like VSCode and right-clicking the function definition and "Find All References", meaning leaking symbols into the output binary is never a reason to do this. The increase in complexity comes from now having a split function definition vs declaration - ie, we'd like to avoid doing this unless there's a compelling reason to do so.

Heck, for this, if we were in C++, we'd be asking you to mark the header definition `constexpr` even

Reply:
Agreed on `constexpr` qualifier for returning 0 for the 7 changed platforms, maybe even `consteval`.
It won't fly for `stlink` and `swlink` because they have a variable (set on `detect_rev()`) and have to to an indirect load to return that.
I am specifically not touching `native` as that's a whole another web of version checks and runtime-conditional behaviour.

Reply:
Right, but for `constexpr` to work (no need for consteval, that serves a different purpose) the functions would need to be put back in the headers. Hence why our wanting to be sure there was a good reason for moving to the implementation TUs. There is: a conflicting existing declaration for the function.